File: app-laravel/src/Shared/Infrastructure/Bus/EventDispatcher.php
<?php

declare(strict_types=1);

namespace Src\Shared\Infrastructure\Bus;

use Src\Shared\Domain\DomainEvent;
use Src\Shared\Domain\IntegrationEvent;
use Src\Shared\Infrastructure\Messaging\RabbitMQPublisher;

/**
 * EVENT DISPATCHER
 * ================
 * Domain Event-ləri dinləyicilərə (listener) çatdırır.
 *
 * İKİ TİP EVENT:
 * 1. Domain Event → Laravel event dispatcher ilə sinxron göndərilir (eyni modul daxilində).
 * 2. Integration Event → RabbitMQ ilə asinxron göndərilir (digər modullara).
 *
 * AXIN:
 * AggregateRoot →pullDomainEvents()→ EventDispatcher
 *   ├─ Domain Event → Laravel Listener (sinxron)
 *   └─ Integration Event → RabbitMQ Publisher (asinxron)
 */
class EventDispatcher
{
    /** @var array<string, callable[]> */
    private array $listeners = [];

    public function __construct(
        private ?RabbitMQPublisher $rabbitMQPublisher = null,
    ) {}

    /**
     * Event-ə listener əlavə et.
     */
    public function listen(string $eventClass, callable $listener): void
    {
        $this->listeners[$eventClass][] = $listener;
    }

    /**
     * Event-ləri dispatch et.
     *
     * @param DomainEvent[] $events
     */
    public function dispatch(array $events): void
    {
        foreach ($events as $event) {
            $this->dispatchSingle($event);
        }
    }

    private function dispatchSingle(DomainEvent $event): void
    {
        $eventClass = get_class($event);

        // 1. Sinxron listener-ləri çağır
        if (isset($this->listeners[$eventClass])) {
            foreach ($this->listeners[$eventClass] as $listener) {
                $listener($event);
            }
        }

        // 2. Integration Event-dirsə, RabbitMQ-ya göndər
        if ($event instanceof IntegrationEvent && $this->rabbitMQPublisher) {
            $this->rabbitMQPublisher->publish($event);
        }
    }
}

File: app-spring/src/main/java/az/ecommerce/shared/infrastructure/bus/EventDispatcher.java
package az.ecommerce.shared.infrastructure.bus;

import az.ecommerce.shared.domain.AggregateRoot;
import az.ecommerce.shared.domain.DomainEvent;
import az.ecommerce.shared.domain.IntegrationEvent;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Component;

/**
 * Laravel: src/Shared/Infrastructure/Bus/EventDispatcher.php
 *
 * Spring-də 2 tip event yayımlayırıq:
 * 1. DomainEvent → Spring ApplicationEventPublisher (sinxron, eyni JVM)
 *    @EventListener-lər sinxron tetiklənir.
 *
 * 2. IntegrationEvent → Spring Modulith Event Publication Registry (Outbox)
 *    avtomatik DB-yə yazılır, sonra RabbitMQ-yə publish olunur.
 *
 * Repository.save()-dən dərhal sonra çağrılır:
 *    eventDispatcher.dispatchAll(aggregate);
 */
@Component
public class EventDispatcher {

    private final ApplicationEventPublisher publisher;

    public EventDispatcher(ApplicationEventPublisher publisher) {
        this.publisher = publisher;
    }

    public void dispatchAll(AggregateRoot aggregate) {
        for (DomainEvent event : aggregate.pullDomainEvents()) {
            dispatch(event);
        }
    }

    public void dispatch(DomainEvent event) {
        // Spring sinxron yayımı (ApplicationEventPublisher)
        publisher.publishEvent(event);

        // IntegrationEvent isə əlavə olaraq Outbox-a yazılacaq
        // (Spring Modulith @ApplicationModuleListener bunu avtomatik edir,
        //  yaxud manual OutboxRelay)
        if (event instanceof IntegrationEvent integrationEvent) {
            publisher.publishEvent(new OutboxEvent(integrationEvent));
        }
    }

    /** Wrapper — OutboxRelay bu event-ləri tutub DB-yə yazır */
    public record OutboxEvent(IntegrationEvent integrationEvent) { }
}
